aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -239,6 +240,159 @@ Tensor wrapped_scalar_tensor(
 
 } // anonymous namespace
 
+// Default imperative implementation of isin(); used when the number of test elements is small.
+static void isin_default(const Tensor& elements, const Tensor& test_elements_flat, bool invert, Tensor& out) {
+  out.fill_(invert);
+  auto iter = TensorIteratorConfig()
+    .set_check_mem_overlap(true)
+    .add_output(out)
+    .add_input(elements)
+    .cast_common_dtype_to_outputs(false)
+    .enforce_safe_casting_to_output(false)
+    .check_all_same_dtype(false)
+    .build();
+  AT_DISPATCH_INTEGRAL_TYPES(elements.scalar_type(), "isin_default", [&]() {
```

Comment:
Why implement isin() using TensorIterator? Can't it be implemented as a composite operation like NumPy's isin() (which delegates to in1d())? I think we have the functions it's composed of.

Reply:
@mruberry, do we want to support in1d()? I somehow have an impression that we are not going to support in1d directly. Maybe I am wrong.

Reply:
in1d() is also a composite operation, though, so we don't need to expose an in1d() function to write isin() as a composite, I think.

It's true we don't want to expose an in1d() function because the docs for in1d() suggest using isin() instead.

Reply:
`isin()` uses `in1d()` in numpy, and we don't want to provide `in1d()`, so, as @glaringlee alluded to, I just implemented a slight variation on `in1d()` as `torch.isin()`.

The [reference implementation iterates](https://github.com/numpy/numpy/blob/fb215c76967739268de71aa4bda55dd1b062bc2e/numpy/lib/arraysetops.py#L578), so I'm not really sure how to get around using `TensorIterator` for this part.

Reply:
That's true that if we want to implement a fast path we may want to write a custom kernel. To that point, how likely are we to follow-up with CUDA support vs a fast path?

Reply:
I think I'm a little confused about what's the fast path here. Are you referring to `isin_default` or `isin_sorting` ? One is faster under certain conditions and the other is faster otherwise. Numpy uses a rough heuristic to decide between the two. The default algo is basically brute force and the sorting-based algo has better time complexity, which is better for larger inputs.

For the default algo, the reference implementation iterates through the test elements, so I'm not sure how this could be written as composite.

For the sorting algo, CUDA support requires stable sort to be available for CUDA- it's not yet, but I think it's planned? cc @glaringlee

Reply:
Sorry, those names are confusing. What I should have said was: 

- is there a composite version of this operation that supports both CPU and CUDA?
- if so, should we implement that first?

Reply:
Cool, I think it's a good idea to try to get at least something implemented for CUDA. I'll probably need some assistance on how to do it, and the default (non-sorting) algo is our best bet since stable sort on CUDA is still further down the road.

Reply:
> I think I'm a little confused about what's the fast path here. Are you referring to `isin_default` or `isin_sorting` ? One is faster under certain conditions and the other is faster otherwise. Numpy uses a rough heuristic to decide between the two. The default algo is basically brute force and the sorting-based algo has better time complexity, which is better for larger inputs.
> 
> For the default algo, the reference implementation iterates through the test elements, so I'm not sure how this could be written as composite.
> 
> For the sorting algo, CUDA support requires stable sort to be available for CUDA- it's not yet, but I think it's planned? cc @glaringlee

GPU stable sort is planed, but not committed. We can either do stable_sort on CPU before mapping it to GPU memory or just implement the non-sorting algo on GPU, probably just implement the non-sorting algo is better option for now.

Reply:
Here's a composite version of the operation:
```cpp
std::vector<int64_t> bc_shape(elements.dim(), 1);
bc_shape.push_back(-1);
out.copy_(invert ? elements.unsqueeze(-1).ne(test_elements.view(bc_shape)).all(-1)
  : elements.unsqueeze(-1).eq(test_elements.view(bc_shape)).any(-1));
```

It works by viewing `test_elements` as the appropriate size to let broadcasting do the rest. It's inefficient and will OOM more easily because it actually materializes tensors of the broadcasted shape.

Any thoughts on ways to do the reduction in a GPU kernel and avoid broadcasting? The issue I'm having is that GPU lambdas cannot capture values by reference, so it can't follow the CPU lambda and capture `test_elements`.